changefeedccl: support azure-event-hub:// for azure kafka streaming

wenyihu6 opened the pull request:
**changefeedccl: support azure-event-hub:// for azure kafka streaming**
Previously, users had to navigate the complexities of obtaining azure event
hub's kafka endpoint and the corresponding sasl_user and sasl_password for azure
streaming.

This patch improves it to support a new scheme azure-event-hub:// with the
following syntax:

```
azure-event-hub://<NamespaceName>.servicebus.windows.net?shared_access_key_name=<KeyName>&shared_access_key=<KeyValue>
```

azure-event-hub:// can now be used to connect to kafka hosted on Azure event
hubs. The sinkURL must include mandatory parameters shared_access_key_name and
shared_access_key.   By default and by requirements, the options
"tls_enabled=true," "sasl_handshake=true," "sasl_enabled=true," and
"sasl_mechanism=PLAIN" are applied, as they are the only supported options.
Other parameters such as "topic_name", and "topic_prefix" are also supported.

Resolves: #103901, #110558

Release note (enterprise change): Changefeeds now support a new scheme
azure-event-hub:// for kafka data streaming to azure event hubs. The sinkURL
must include mandatory parameters shared_access_key_name and shared_access_key.
By default and by requirements, the options "tls_enabled=true,"
"sasl_handshake=true," "sasl_enabled=true," and "sasl_mechanism=PLAIN" are
applied, as they are the only supported options. Other parameters such as
"topic_name", and "topic_prefix" are also supported.

An example URI is:
```
azure-event-hub://myeventhubs.servicebus.windows.net:9093?shared_access_key_name=abc&shared_access_key=123
```
---
**changefeedccl: redact shared_access_key from SHOW JOBS**

Now that we support azure-event-hub:// scheme, this patch redacts
shared_access_key info for azure event hub from SHOW JOBS output.
See also: https://github.com/cockroachdb/cockroach/issues/103901
Epic: None
Release note: None

---
**roachtest/cdc: add roachtest for azure-event-hub://**

This patch adds a roachtest for kafka streaming using azure-event-hub://.

wenyihu6 committed:
changefeedccl: support azure-event-hub:// for azure kafka streaming

Previously, users had to navigate the complexities of obtaining azure event
hub's kafka endpoint and the corresponding sasl_user and sasl_password for azure
streaming.

This patch improves it to support a new scheme azure-event-hub:// with the
following syntax:

```
azure-event-hub://<NamespaceName>.servicebus.windows.net?shared_access_key_name=<KeyName>&shared_access_key=<KeyValue>
```

azure-event-hub:// can now be used to connect to kafka hosted on Azure event
hubs. The sinkURL must include mandatory parameters shared_access_key_name and
shared_access_key.   By default and by requirements, the options
"tls_enabled=true," "sasl_handshake=true," "sasl_enabled=true," and
"sasl_mechanism=PLAIN" are applied, as they are the only supported options.
Other parameters such as "topic_name", and "topic_prefix" are also supported.

Resolves: #103901, #110558

Release note (enterprise change): Changefeeds now support a new scheme
azure-event-hub:// for kafka data streaming to azure event hubs. The sinkURL
must include mandatory parameters shared_access_key_name and shared_access_key.
By default and by requirements, the options "tls_enabled=true,"
"sasl_handshake=true," "sasl_enabled=true," and "sasl_mechanism=PLAIN" are
applied, as they are the only supported options. Other parameters such as
"topic_name", and "topic_prefix" are also supported.

An example URI is:
```
azure-event-hub://myeventhubs.servicebus.windows.net:9093?shared_access_key_name=abc&shared_access_key=123
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/115806)
<!-- Reviewable:end -->

jayshrivastava reviewed (changes requested):
Looks pretty good. Minor comments!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-:-Nl9d_d_7rnJ3qUV2OVp:bxng4fu)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @wenyihu6)

___
*[`-- commits` line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9cV5TADsQ-t3J8b4o):*
We probably don't need  "insecure_tls_skip_verify". I added that for confluent as a hack for roachtests. Will likely remove that during this release. 
___
*[`-- commits` line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9czDG9GsIgOBk3rdc):*
typo. should be "myeventhubs"? 
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 992 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9_FVH2xeSlV-eeGV_:-Nl9_FVH2xeSlV-eeGVa:b-nay4zv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka.go#L992)):*
> ```Go
> // newMissingParameterError returns an error message for using unsupported
> // values in sinkURL.
> func newRequiredValueError(param string, unsupportedValue, allowedValue string) error {
> ```

Can you rename this to `NewUnsupportedValueForParameterErrors`?
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 1018 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9_4FLDozo2eJpMzi5:-Nl9_4FLDozo2eJpMzi6:bqxtxzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka.go#L1018)):*
> ```Go
> 		}
> 		if wasSet && dest != allowedBoolValue {
> 			return errors.Newf("unsupported value %t for parameter %s, please use %t instead", dest,
> ```

This can use `newRequiredValueError` right?
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 1129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9bzCQ5QFNk3uAslR7:-Nl9bzCQ5QFNk3uAslR8:bhezrqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka.go#L1129)):*
> ```Go
> // parameters from the sinkURL. Additionally, it validates options based on the
> // given sinkURL and returns an error for unsupported values.
> func extendKafkaConfigWithDefault(u *sinkURL, dialConfig kafkaDialConfig) (kafkaDialConfig, error) {
> ```

I think this can be renamed to "setDefaultParametersForConfluentAndAzure". I would add a comment to say exactly which parameters its setting to true. 
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 1177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9cO8V9MK2j3fywTcC:-Nl9cO8V9MK2j3fywTcD:bvrynes) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka.go#L1177)):*
> ```Go
> // https://learn.microsoft.com/en-us/azure/event-hubs/azure-event-hubs-kafka-overview
> // on how to connect to azure event hub kafka protocol.
> func buildAzureKafkaConfig(u sinkURL) (dialConfig kafkaDialConfig, _ error) {
> ```

Can you please add some tests for params like"topic_name" and "topic_prefix" (and whichever else non-auth-related params we want to allow).
___
*[`pkg/ccl/changefeedccl/sink_kafka_connection_test.go` line 371 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9dzpr2Cd0gqBtjnvn:-Nl9dzpr2Cd0gqBtjnvo:b-acd307) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka_connection_test.go#L371)):*
> ```Go
> 			expectedConfig, expectedError := buildDialConfig(sinkURL{URL: oldUri})
> 			actualConfig, actualError := buildDialConfig(sinkURL{URL: newUri})
> 			require.Equal(t, expectedError, actualError)
> ```

We should store the dial cfg structs that you expect to see in the test cases above. I think testing this way is a bit risky. 
___
*[`pkg/ccl/changefeedccl/sink_kafka_connection_test.go` line 375 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9ddIK3DJgUzScZQAE:-Nl9ddIK3DJgUzScZQAF:byuto68) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka_connection_test.go#L375)):*
> ```Go
> 		})
> 	}
> }
> ```

Would you mind running the `kafka-auth` roachtest to make sure it passes. Just bc you made changes to confluent, there's a small chance it could fail. Easier to check now than have it fail in nightlies. 


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
changefeedccl: redact shared_access_key from SHOW JOBS

Now that we support azure-event-hub:// scheme, this patch redacts
shared_access_key info for azure event hub from SHOW JOBS output.
See also: #103901
Epic: None
Release note: None

wenyihu6 commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -1021,6 +1021,10 @@ func runCDCKafkaAuth(ctx context.Context, t test.Test, c cluster.Cluster) {
 		},
+		{
+			"create changefeed with azure-event-hub scheme",
+			fmt.Sprintf("%sshared_access_key_name=plain&shared_access_key=plain-secret", kafka.sinkURLAsAzureEventHubUrl(ctx)),
```
I will come back with a better answer when things are clearer - I'm still testing this roachtest out.

wenyihu6 replied:
The comment no longer applied. We addressed the comment in the other PR https://github.com/cockroachdb/cockroach/pull/116950.

wenyihu6 reviewed (commented):
Manually confirmed that this works, but adding azure authentication to cdc/kafka-auth is still blocked. Opened an [issue](https://github.com/cockroachdb/cockroach/issues/116323) to track this - working on it now. Do we want to defer merging this PR until we have the roachtest piece figured out or should we merge this first?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-:-NlYw3Z58HA2mvzTtF04:bqk1y9r)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @jayshrivastava, @miretskiy, and @srosenberg)

___
*[`-- commits` line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9cV5TADsQ-t3J8b4o):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

We probably don't need  "insecure_tls_skip_verify". I added that for confluent as a hack for roachtests. Will likely remove that during this release.
</blockquote></details>

Done.
___
*[`-- commits` line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9czDG9GsIgOBk3rdc):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

typo. should be "myeventhubs"?
</blockquote></details>

Oops. Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 992 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9_FVH2xeSlV-eeGV_:-NlYuliK2oOFfZ7SCG81:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka.go#L992)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Can you rename this to `NewUnsupportedValueForParameterErrors`?
</blockquote></details>

Done. 
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 1018 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9_4FLDozo2eJpMzi5:-NlYuoG_D4v2rdCp7ICO:bh2drzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka.go#L1018)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

This can use `newRequiredValueError` right?
</blockquote></details>

Ahh yeah done. 
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 1129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9bzCQ5QFNk3uAslR7:-NlYutsR72hjWYKALdy2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka.go#L1129)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

I think this can be renamed to "setDefaultParametersForConfluentAndAzure". I would add a comment to say exactly which parameters its setting to true.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 1177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9cO8V9MK2j3fywTcC:-NlYuvHm2eHxnZmVHGpr:b-6joxa6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka.go#L1177)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Can you please add some tests for params like"topic_name" and "topic_prefix" (and whichever else non-auth-related params we want to allow).
</blockquote></details>

I added more tests in `TestChangefeedExternalConnections` for this purpose. Wasn't able to add more tests for this in `TestBuildAzureKafkaConfig` since topic_name and topic_prefix are consumed before buildKafkaConfig was called. 
___
*[`pkg/ccl/changefeedccl/sink_kafka_connection_test.go` line 371 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9dzpr2Cd0gqBtjnvn:-NlYvTzu97SDyIvlZVq9:b7chb9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka_connection_test.go#L371)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

We should store the dial cfg structs that you expect to see in the test cases above. I think testing this way is a bit risky.
</blockquote></details>

Ahh true done. 
___
*[`pkg/ccl/changefeedccl/sink_kafka_connection_test.go` line 375 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/115806#-Nl9ddIK3DJgUzScZQAE:-NlYvY9H9s8X39373OJ0:bqg6es3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b91a03c0cc4aa7af982c051413b9511a12065fb/pkg/ccl/changefeedccl/sink_kafka_connection_test.go#L375)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Would you mind running the `kafka-auth` roachtest to make sure it passes. Just bc you made changes to confluent, there's a small chance it could fail. Easier to check now than have it fail in nightlies.
</blockquote></details>

yup - confirmed it's working 
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1026 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/115806#gh-1423220856:-NlYvoNS8mOkkmBn-51_:bz04enu) ([raw file](https://github.com/cockroachdb/cockroach/blob/54efa8e31cf80ffa16d0c39a7760a3e9ad09fdb0/pkg/cmd/roachtest/tests/cdc.go#L1026)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

I will come back with a better answer when things are clearer - I'm still testing this roachtest out. 
</blockquote></details>

I'm putting roachtest piece on hold until we figure out what to do with [#105576](https://github.com/cockroachdb/cockroach/pull/105576).


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Feel free to hold PR review off until we merge https://github.com/cockroachdb/cockroach/pull/116950

wenyihu6 committed:
roachtest/cdc: add roachtest for azure-event-hub://

This patch changes roachtest kafka-azure for azure event hub using the new
scheme azure-event-hub://.

wenyihu6 commented:
Confirmed that the test is working -https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAzureBazel/13585661?buildTab=overview&hideProblemsFromDependencies=false&hideTestsFromDependencies=false

wenyihu6 commented:
The first commit is from https://github.com/cockroachdb/cockroach/pull/117893.

wenyihu6 commented:
Friendly ping on the review : )

jayshrivastava commented on pkg/ccl/changefeedccl/show_changefeed_jobs_test.go:
```diff
@@ -170,8 +174,8 @@ func TestShowChangefeedJobsRedacted(t *testing.T) {
 			expectedSinkURI = strings.Replace(expectedSinkURI, certSecret, "redacted", 1)
 			expectedDescription := strings.Replace(createStmt, apiSecret, "redacted", 1)
 			expectedDescription = strings.Replace(expectedDescription, certSecret, "redacted", 1)
```
expectedDescription is written twice here.

wenyihu6 replied:
I will leave it as it is for now : )

wenyihu6 replied:
I went with the second option here. Lmk if you are against it.

jayshrivastava reviewed (approved):
LGTM!

wenyihu6 commented:
TFTRs!

bors r=jayshrivastava

wenyihu6 commented:
TFTRs!

bors r=jayshrivastava
